_posts/2021-09-26-where-the-store-is-stored.md

```diff
@@ -0,0 +1,292 @@
+
+In this post, we're focusing on "store" kind of questions.
+
+# The multiple places we can store our stores
+
+The reality of the store that lives "somewhere" in the code is the very
+problem. The store in itself, whether a simple variable or a full Redux store,
+has be stored somewhere in our browser, NodeJS or other fancy javascript
+engine runtime. 
+
+Deciding *where our stores are stored* is not trivial and comes with big
+impacts. Let me draw here the three main potential locations we can use for our
+stores.
+
+## Modules
+
+We can use modules to host our stores. It's quite convenient and it makes full
+use of the Javascript ES6 language.
+
+Let's say we have a module storing Cars, our state here.
+
+```js
+// cars/store.js
+
+const _cars = []
+
+export function add(car) {
+  _cars.push(car);
+}
+
+export function list() {
+  return _cars;
+} 
+```
+
+This `cars/store.js` module acts as a very naive store with both *read*
+(`list`) and *write* (`add`) functionalities.
+
+Using this store in the rest of the codebase is very easy: we just have to use
+classic ES6 imports.
+
+```js
+// another/part/of/the/codebase.js
+import * as Cars from '../path/to/cars/store';
+
+function doingSomething() {
+  console.log('current cars: ', Cars.list());
+
+  const newCar = "My old car";
+
+  cars.add(newCar);
+}
+```
+
+And *voilà*! Because ES6 module are statefull, the `_cars` "private" variable
+will stay the same across all imports, making `cars/store.js` a store by itself.
+
+Should we need a reactive store (ie. being able to *subscribe* to state
+changes), we would write our own subscribe function by hand or use classic store
+librairies (some opinionated like Redux, Mobx, etc, some more transversal like
+the [@fp-51/store](https://github.com/fp51/store-library) I authored one or
+two years ago). 
+
+## Librairies / Framework location
+
+Another very common way of storing our store is using the locations provided by
+the main view library or framework we use (React, Vue, etc.).
+
+Let's say that we're using React here. The *go to* store location is then a
+simple `useState` hook.
+
+```js
+// app.js
+import React from 'react';
+
+import { UserContext } from './cars.js'
+
+function App({ children }) {
+  const [cars, setCars] = React.useState([]);
+
+  const addCar = car => setCars(cars => [...cars, car]);
+
+  return (
+    <MyComponent cars={cars} addCar={addCar} />
+  )
+}
+```
+
+We're not using modules to host our store anymore. React does it for us via
+`useState`.  Note that we're reactive by default here: calling `addCar` in
+`MyComponent` will rerender all `<App />`.
+
+Using `useState` alone usually doesn't scale well as we're forced to pass `cars`
+and `setCar` everywhere through props. The issue is called [props
+drilling](https://sebhastian.com/react-prop-drilling/). To overcome this, we
+can add a `Context` to make our state and its API available to every child
+components via a simple `useContext` - even if [it's not made for
```

Comment:
Funnily, SwiftUI solves this the same way with `@EnvironmentObject`, but they made it clear it was the purpose.

Reply:
ah interesting !